apiserver/facades/client/modelupgrader/controllerupgrader.go

```diff
@@ -0,0 +1,278 @@
+		).Add(coreerrors.NotValid)
+		return result, errors.Capture(e)
+	case errors.Is(err, modelagenterrors.AgentStreamNotValid):
+		e := errors.Errorf(
+			"agent stream %q is not recognised as a valid value", arg.AgentStream,
+		).Add(coreerrors.NotValid)
+		return result, errors.Capture(e)
+	case err != nil:
+		return result, errors.Capture(apiservererrors.ServerError(err))
+	}
+
+	result.ChosenVersion = targetVersion
+	return result, nil
+}
+
+// doUpgrade has the responsibility of delegating the upgrade to the service. It determines which func to invoke
+// by interrogating the values set in [params.UpgradeModelParams].
+// A post-processing step is performed to map the errors returned from the service to ones the existing API
+// conforms to.
+func (c *ControllerUpgraderAPI) doUpgrade(ctx context.Context, arg params.UpgradeModelParams) (params.UpgradeModelResult, error) {
+	var (
+		hasStreamChange  = arg.AgentStream != ""
+		hasTargetVersion = arg.TargetVersion != semversion.Zero
+		targetStream     modelagent.AgentStream
+		targetVersion    = arg.TargetVersion
+		upgrader         func(context.Context) error
+		dryRunValidate   func(context.Context) (semversion.Number, error)
+		result           params.UpgradeModelResult
+		err              error
+	)
+
+	// Parse the agent stream.
+	if arg.AgentStream != "" {
+		targetStream, err = modelagent.AgentStreamFromCoreAgentStream(agentbinary.AgentStream(arg.AgentStream))
+		if err != nil {
+			if errors.Is(err, coreerrors.NotValid) {
+				return result, errors.Capture(errors.Errorf(
+					"agent stream %q is not recognised as a valid value", arg.AgentStream,
+				).Add(coreerrors.NotValid))
+			}
+			return result, errors.Capture(err)
+		}
+	}
+
+	// Delegate it to the service depending on what arguments
+	// are supplied.
+	switch {
+	case hasTargetVersion && hasStreamChange:
+		upgrader = func(ctx context.Context) error {
+			return c.upgraderService.UpgradeControllerToVersionAndStream(ctx, targetVersion, targetStream)
+		}
+		dryRunValidate = func(ctx context.Context) (semversion.Number, error) {
+			return targetVersion, c.upgraderService.CanUpgradeControllerToVersionWithStream(ctx, targetVersion, targetStream)
+		}
+	case hasTargetVersion && !hasStreamChange:
+		upgrader = func(ctx context.Context) error {
+			return c.upgraderService.UpgradeControllerToVersion(ctx, targetVersion)
+		}
+		dryRunValidate = func(ctx context.Context) (semversion.Number, error) {
+			return targetVersion, c.upgraderService.CanUpgradeControllerToVersion(ctx, targetVersion)
+		}
+	case !hasTargetVersion && hasStreamChange:
+		upgrader = func(ctx context.Context) error {
+			version, err := c.upgraderService.UpgradeControllerWithStream(ctx, targetStream)
+			targetVersion = version
+			return err
+		}
+		dryRunValidate = func(ctx context.Context) (semversion.Number, error) {
+			return c.upgraderService.CanUpgradeControllerWithStream(ctx, targetStream)
+		}
+	case !hasTargetVersion && !hasStreamChange:
```

Comment:
I would make this case the "default", I would do this because it is the zero value of bool and also it makes sure your func variables are set with something. Without this there could be a programmatic error later on that would result in this code trying to use the nil func variables resulting in a panic .